### PR TITLE
Implement Container methods

### PR DESCRIPTION
**Pull request by Limeth**
[SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1209/) | [SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1491)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 1]
```diff
@@ -837,4 +839,28 @@ public void onSetContainer(IInventory chestInventory, CallbackInfo ci) {
             ((IMixinContainer) this.openContainer).setSpectatorChest(true);
         }
     }
+
+    @Override
+    public void openContainer(Container container) {
+        if (openContainer != inventoryContainer) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 2]
```diff
@@ -837,4 +839,28 @@ public void onSetContainer(IInventory chestInventory, CallbackInfo ci) {
...
+        this.getNextWindowId();
+        this.connection.sendPacket(new SPacketOpenWindow(this.currentWindowId, guiId, containerInventory.getDisplayName(), containerInventory.getSizeInventory()));
+        this.openContainer = (net.minecraft.inventory.Container) container;
+        this.openContainer.windowId = this.currentWindowId;
+        this.openContainer.addListener(this$);
+    }
+
+    @Override
+    public void closeContainer(Container container) {
+        IMixinContainer mixinContainer = (IMixinContainer) openContainer;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java [hunk 3]
```diff
@@ -274,5 +285,47 @@ public void setSpectatorChest(boolean spectatorChest) {
...
     }
+
+    @Override
+    public Set<Player> getViewers() {
+        return ImmutableSet.copyOf((Set) playerList);
+    }
+
+    @Override
+    public boolean hasViewers() {
+        return playerList.size() > 0;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java [hunk 4]
```diff
@@ -274,5 +285,47 @@ public void setSpectatorChest(boolean spectatorChest) {
     public void setCanInteractWith(@Nullable Predicate<EntityPlayer> predicate) {
         this.canInteractWithPredicate = Optional.ofNullable(predicate); // TODO mixin into all classes extending container
     }
+
+    @Override
+    public Set<Player> getViewers() {
+        return ImmutableSet.copyOf((Set) playerList);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainer.java]: `this.`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java [hunk 5]
```diff
@@ -837,4 +839,28 @@ public void onSetContainer(IInventory chestInventory, CallbackInfo ci) {
...
+        }
+
+        IInventory containerInventory = ((MinecraftInventoryAdapter) container).getInventory().get(0);
+        String guiId = containerInventory instanceof IInteractionObject ? ((IInteractionObject) containerInventory).getGuiID() : "minecraft:chest";
+
+        getNextWindowId();
+        this.connection.sendPacket(new SPacketOpenWindow(this.currentWindowId, guiId, containerInventory.getDisplayName(), containerInventory.getSizeInventory()));
+        this.openContainer = (net.minecraft.inventory.Container) container;
+        this.openContainer.windowId = this.currentWindowId;
+        this.openContainer.addListener(this$);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: `this.this$`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java (hunk 5 above)
Comment (Limeth) [reply src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayerMP.java]: I wasn't sure about that one. :smile:

**Comment by parlough** [comment]
Going to close this due to inactivity, if you'd still like to get this in, update for the latest changes and open a new pull request.

Thanks for your contribution!

**Pull request closed without merging**